### PR TITLE
Add cardinality calculation for Dataset.unbatch() when possible

### LINKED ISSUE #39136: Dataset.unbatch() sets cardinality to -2 even when batch size is known
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 x64 (1909)
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de410 2.1.0
- Python version: 3.7.7
- CUDA/cuDNN version: Using CPU
- GPU model and memory: Using CPU

**Describe the current behavior**
Doing `Dataset.unbatch()` on dataset with known batch size resets cardinality to -2 (unknown).

**Describe the expected behavior**
When batch size of dataset is known, it should set cardinality to `batch_size * cardinality`.

**Standalone code to reproduce the issue**
```
import tensorflow as tf
ds = tf.data.Dataset.range(10) # shape=()
ds = ds.batch(2, drop_remainder=True) # shape=(2,)
print(tf.data.experimental.cardinality(ds)) # 5
ds = ds.unbatch() # shape=()
print(tf.data.experimental.cardinality(ds)) # Should be 10, but is -2 (unknown)
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.
Although cardinality is currently experimental, it is used when traning keras model.

### PR DESCRIPTION
**Pull request by yongtang**
This PR tries to address the issue raised in #39136 where cardinality
of Dataset.unbatch() was always UNKNOWN, even if it might be known
in certain situations.

This PR add the cardinality calculation in case the input cardinality
is known and the leading dim of the output shape is known.

This PR fixes #39136.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc [hunk 1]
```diff
@@ -38,8 +38,14 @@ class UnbatchDatasetOp : public UnaryDatasetOpKernel {
     explicit Dataset(OpKernelContext* ctx, DatasetBase* input)
         : DatasetBase(DatasetContext(ctx)), input_(input) {
       input_->Ref();
+      known_batch_size_ = -1;
       for (const PartialTensorShape& shape : input->output_shapes()) {
         if (!shape.unknown_rank()) {
+          if (known_batch_size_ < 0) {
+            if (shape.dim_size(0) >= 0) {
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc:45]: combine the two `if` clauses:
```
if (known_batch_size_ < 0 && shape.dim_size(0) >= 0) {
  ...
}
```

File: tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc [hunk 2]
```diff
@@ -222,6 +239,7 @@ class UnbatchDatasetOp : public UnaryDatasetOpKernel {
 
     const DatasetBase* const input_;
     std::vector<PartialTensorShape> shapes_;
+    int64 known_batch_size_;
```
Comment (aaudiber) [inline tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc:242]: can we call this just `batch_size_`? Then add a comment that it may or may not be known, with -1 representing unknown.

File: tensorflow/python/data/experimental/kernel_tests/cardinality_test.py [hunk 3]
```diff
@@ -134,6 +134,15 @@ def _test_combinations():
...
        cardinality.UNKNOWN),
       ("Take4", lambda: dataset_ops.Dataset.range(5).repeat().take(2), 2),
+      ("Unbatch1",
+       lambda: dataset_ops.Dataset.range(5).batch(2, drop_remainder=True).unbatch(), 4),
+      ("Unbatch2",
+       lambda: dataset_ops.Dataset.range(5).batch(2, drop_remainder=False).unbatch(), cardinality.UNKNOWN),
+      ("Unbatch3",
+       lambda: dataset_ops.Dataset.range(5).batch(2, drop_remainder=True).filter(lambda _: True).unbatch(),
+       cardinality.UNKNOWN),
+      ("Unbatch4", lambda: dataset_ops.Dataset.range(5).batch(2, drop_remainder=True).repeat().unbatch(),
```
Comment (aaudiber) [inline tensorflow/python/data/experimental/kernel_tests/cardinality_test.py:144]: Add test with 2 components, where only the second component's batch size is known:
```
lambda: dataset_ops.Dataset.zip(
  dataset_ops.Dataset.range(4).batch(2, drop_remainder=False),
  dataset_ops.Dataset.range(5).batch(2, drop_remainder=True))
```

**Review by aaudiber — CHANGES_REQUESTED** [review]
Thanks for implementing this @yongtang! Looks great overall, I just left a few minor comments.

File: tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc (hunk 1 above)
Comment (yongtang) [reply tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc]: Thanks @aaudiber, done.

File: tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc (hunk 2 above)
Comment (yongtang) [reply tensorflow/core/kernels/data/experimental/unbatch_dataset_op.cc]: Thanks @aaudiber, updated.

File: tensorflow/python/data/experimental/kernel_tests/cardinality_test.py (hunk 3 above)
Comment (yongtang) [reply tensorflow/python/data/experimental/kernel_tests/cardinality_test.py:144]: Done.

**Comment by yongtang** [comment]
Thanks @aaudiber for the review, the PR has been updated. Please take a look and let me know if there are any other issues.

**Review by aaudiber — APPROVED** [review]
